flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java

```diff
@@ -959,24 +947,44 @@ public void testKryoRegisteringRestoreResilienceWithDefaultSerializer() throws E
-            expectedException.expect(
-                    anyOf(
-                            isA(ExpectedKryoTestException.class),
-                            Matchers.<Throwable>hasProperty(
-                                    "cause", isA(ExpectedKryoTestException.class))));
-
-            // state backends that eagerly deserializes (such as the memory state backend) will fail
-            // here
-            backend = restoreKeyedBackend(IntSerializer.INSTANCE, snapshot2, env);
-
-            state =
-                    backend.getPartitionedState(
-                            VoidNamespace.INSTANCE, VoidNamespaceSerializer.INSTANCE, kvId);
-
-            backend.setCurrentKey(1);
-            // state backends that lazily deserializes (such as RocksDB) will fail here
-            state.value();
-
+            assertThatThrownBy(
+                            () -> {
+                                // state backends that eagerly deserializes (such as the memory
+                                // state backend) will fail
+                                // here
+                                CheckpointableKeyedStateBackend<Integer> restoreBackend = null;
+                                try {
+                                    restoreBackend =
+                                            restoreKeyedBackend(
+                                                    IntSerializer.INSTANCE, snapshot2, env);
+
+                                    ValueState<TestPojo> restoreState =
+                                            restoreBackend.getPartitionedState(
+                                                    VoidNamespace.INSTANCE,
+                                                    VoidNamespaceSerializer.INSTANCE,
+                                                    new ValueStateDescriptor<>("id", pojoType));
+
+                                    restoreBackend.setCurrentKey(1);
+                                    // state backends that lazily deserializes (such as RocksDB)
+                                    // will fail here
+                                    restoreState.value();
+
+                                    restoreBackend.dispose();
+                                } finally {
+                                    if (restoreBackend != null) {
+                                        IOUtils.closeQuietly(restoreBackend);
+                                        restoreBackend.dispose();
+                                    }
+                                }
+                            })
+                    .satisfiesAnyOf(
+                            actual ->
+                                    assertThat(actual)
+                                            .isInstanceOf(ExpectedKryoTestException.class),
+                            actual ->
+                                    assertThat(actual)
+                                            .hasFieldOrProperty("cause")
```

Comment:
Could you please clarify why should we check it?
In my opinion `Throwable` has `cause` field and this check will always pass for anything extending `Throwable`

Reply:
This is because some catch blocks wrap the inner exception, i.e., `ExpectedKryoTestException` (thrown by `CustomKryoTestSerializer`).

For example, `HeapKeyedStateBackendBuilder#restoreState` wrap `ExpectedKryoTestException` in `BackendBuildingException`.

Thus, we need to check the `cause` field of the thrown exception.

Reply:
there is `.hasCauseInstanceOf` method in assertJ e.g.
```java
assertThatThrownBy(() -> ...)
.hasCauseInstanceOf(ExpectedCauseException.class);
```
will not it help here?

Reply:
Yes, I think `hasCauseInstanceOf()` is sufficient for the assertion, and `.hasFieldOrProperty("cause")` can be removed. The assertion code will become:

```
.satisfiesAnyOf(
        actual ->
                assertThat(actual)
                        .isInstanceOf(ExpectedKryoTestException.class),
        actual ->
                assertThat(actual)
                        .hasCauseInstanceOf(ExpectedKryoTestException.class));
```